FWCore/Integration/plugins/AcquireIntESProducer.cc

```diff
@@ -0,0 +1,152 @@
+// -*- C++ -*-
+//
+// Package:    FWCore/Integration
+// Class:      AcquireIntESProducer
+//
+/**\class edmtest::AcquireIntESProducer
+
+  Description: Used in tests of the asynchronous ESProducer.
+*/
+// Original Author:  W. David Dagenhart
+//         Created:  12 January 2023
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/ESProducerExternalWork.h"
+#include "FWCore/Framework/interface/EventSetupRecordKey.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/Framework/interface/ValidityInterval.h"
+#include "FWCore/Integration/interface/ESTestData.h"
+#include "FWCore/Integration/interface/ESTestRecords.h"
+#include "FWCore/Integration/interface/IOVTestInfo.h"
+#include "WaitingServer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "FWCore/Utilities/interface/ESInputTag.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include <memory>
+#include <unistd.h>
+#include <vector>
+
+namespace edmtest {
+
+  class AcquireIntESProducer : public edm::ESProducerExternalWork {
+  public:
+    AcquireIntESProducer(edm::ParameterSet const&);
+
+    ~AcquireIntESProducer() override;
+
+    void initConcurrentIOVs(EventSetupRecordKey const&, unsigned int nConcurrentIOVs) override;
+
+    void acquire(ESTestRecordI const&, edm::WaitingTaskWithArenaHolder);
+
+    std::unique_ptr<ESTestDataI> produce(ESTestRecordI const&);
+
+    std::unique_ptr<ESTestDataB> produceESTestDataB(ESTestRecordB const&);
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::ESGetToken<IOVTestInfo, ESTestRecordI> token_;
+    std::vector<test_acquire::Cache> caches_;
+    std::unique_ptr<test_acquire::WaitingServer> server_;
+    const unsigned int numberOfIOVsToAccumulate_;
+    const unsigned int secondsToWaitForWork_;
+  };
+
+  AcquireIntESProducer::AcquireIntESProducer(edm::ParameterSet const& pset)
+      : numberOfIOVsToAccumulate_(pset.getUntrackedParameter<unsigned int>("numberOfIOVsToAccumulate")),
+        secondsToWaitForWork_(pset.getUntrackedParameter<unsigned int>("secondsToWaitForWork")) {
+    auto collector = setWhatAcquiredProduced(this, "fromAcquireIntESProducer");
+    setWhatProduced(this, &edmtest::AcquireIntESProducer::produceESTestDataB);
+    token_ = collector.consumes<IOVTestInfo>(edm::ESInputTag{"", ""});
```

Comment:
I would reverse these two lines
```suggestion
    token_ = collector.consumes<IOVTestInfo>(edm::ESInputTag{"", ""});
    setWhatProduced(this, &edmtest::AcquireIntESProducer::produceESTestDataB);
```
because the `token_` is related to the functions registered by `setWhatAcquiredProduced()`.

Reply:
Done. Better.